### DISCUSSION TITLE
Confusion about `PipeReader`

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#81819] [Q&A] [answered]
According to https://learn.microsoft.com/en-us/dotnet/standard/io/pipelines#pipereader:

> [PipeReader](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader) manages memory on the caller's behalf. Always call [PipeReader.AdvanceTo](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader.advanceto) after calling [PipeReader.ReadAsync](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader.readasync). This lets the `PipeReader` know when the caller is done with the memory so that it can be tracked. The `ReadOnlySequence<byte>` returned from `PipeReader.ReadAsync` is only valid until the call the `PipeReader.AdvanceTo`. It's illegal to use `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo`.

If I want to reference a `ReadOnlyMemory<byte>` (which inside `ReadOnlySequence<byte>`) in memory for a long time, should I `ReadOnlySequence<byte>.CopyTo`/`SequenceReader<byte>.TryCopyTo`? If I use `SequenceMarshal.TryGetReadOnlyMemory`, are there any potential problems?

**Comment by AlgorithmsAreCool** [comment, marked as answer]
Not sure if you have gotten your answer elsewhere, but if you want to keep the data after you call `AdvanceTo` you will need to copy it. If you use `TryGetReadOnlyMemory` it will give you a reference to the underlying memory that the Pipe owns. After `AdvanceTo` is called then the Pipe may reuse or clobber that memory.

The alternative of course is to delay the call to `AdvanceTo` until you are done with the memory, but that might not be possible as you are hinting.
